libafl/src/stages/pruning.rs

```diff
@@ -0,0 +1,151 @@
+//! Corpus pruning stage
+
+use core::marker::PhantomData;
+
+use libafl_bolts::{rands::Rand, Error};
+
+#[cfg(feature = "std")]
+use crate::{events::EventRestarter, state::Stoppable};
+use crate::{
+    corpus::Corpus,
+    stages::Stage,
+    state::{HasCorpus, HasRand, UsesState},
+};
+
+#[derive(Debug)]
+/// The stage to probablistically disable a corpus entry.
+/// This stage should be wrapped in a if stage and run only when the fuzzer perform restarting
+/// The idea comes from `https://mschloegel.me/paper/schiller2023fuzzerrestarts.pdf`
+pub struct CorpusPruning<EM> {
+    /// The chance of retaining this corpus
+    prob: f64,
+    phantom: PhantomData<EM>,
+}
+
+impl<EM> CorpusPruning<EM> {
+    fn new(prob: f64) -> Self {
+        Self {
+            prob,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<EM> Default for CorpusPruning<EM> {
+    fn default() -> Self {
+        Self::new(0.05)
+    }
+}
+
+impl<EM> UsesState for CorpusPruning<EM>
+where
+    EM: UsesState,
+{
+    type State = EM::State;
+}
+
+impl<E, EM, Z> Stage<E, EM, Z> for CorpusPruning<EM>
+where
+    EM: UsesState,
+    E: UsesState<State = Self::State>,
+    Z: UsesState<State = Self::State>,
+    Self::State: HasCorpus + HasRand,
+{
+    #[allow(clippy::cast_precision_loss)]
+    fn perform(
+        &mut self,
+        _fuzzer: &mut Z,
+        _executor: &mut E,
+        state: &mut Self::State,
+        _manager: &mut EM,
+    ) -> Result<(), Error> {
+        // Iterate over every corpus entry
+        let n_corpus = state.corpus().count_all();
```

Comment:
Why don't you just call `corpus().next()` in a loop? Because you also want disabled testcases? Do we need an extra API?
https://github.com/AFLplusplus/LibAFL/blob/d8e53d5b657b6e8cc6e06a13c81923cd59b2833a/libafl/src/corpus/mod.rs#L132

Reply:
Why do you even want disabled testcases?

Reply:
> Because you also want disabled testcases? 

yes

> Do we need an extra API?

maybe yes

> Why do you even want disabled testcases?

Because the point of this corpus pruning is to hide some inputs so that fuzzer won't schedule it. but we don't discard it because we would reactivate them again later

Reply:
But you don't want need to disable disabled testcases, right now you do(?)

Reply:
@tokatoka